Spec deprecatedURNtoURL

JensenPaul opened the pull request:
Addresses #160


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/JensenPaul/fenced-frame/pull/161.html" title="Last updated on Jun 6, 2024, 1:33 PM UTC (6495120)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/161/4966a9e...JensenPaul:6495120.html" title="Last updated on Jun 6, 2024, 1:33 PM UTC (6495120)">Diff</a>

qingxinwu commented on spec.bs:
```diff
@@ -1437,12 +1437,20 @@ the API {{Window/navigator}}.{{Navigator/deprecatedReplaceInURN()}}, which allow
 macros into the [=fenced frame config/mapped url=] corresponding to a given [=urn uuid=] or
 {{FencedFrameConfig}}.
 
+Note: To help with ease of adoption,
```
Does this one have a different plan from deprecatedReplaceInURN()? The note above says until 2026, and here it says third party cookie deprecation. If they have the same plan, we should probably update both, or possibly a single note paragraph to cover both the two APIs?

JensenPaul replied:
They're different timelines, hence the different wording and different links.

qingxinwu commented:
LGTM

domfarolino commented:
I saw this in my inbox but I noticed I was not added as a reviewer. Should I take a look? Feel free to add me as a reviewer and I will take a look after I'm back from vacation.

JensenPaul commented on spec.bs:
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
+     1. Let |mapping| be |global|'s [=associated Document=]'s [=node navigable=]'s
+        [=navigable/traversable navigable=]'s [=traversable navigable/fenced frame config mapping=].
+
+     1. Let |config| be the result of [=fenced frame config mapping/finding a config=] in |mapping|
```
Does that mean callers have to wait for a Protected Audience auction to complete before calling `deprecatedURNtoURL()`?
It does look like we require the mapping to be finalized: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/ad_auction_service_impl.cc;l=487;drc=2b8b8c15ea96feadd34e04ca947e5f05069e3b24
Should I just change this to instead say this?
```
    1. If |mapping|'s [=fenced frame config mapping/finalized config mapping=][|urn|], does not
       [=map/exist=], [=queue a global task=] on the [=DOM manipulation task source=]
       given |global|, to [=resolve=] |p| with {{undefined}}, and abort these steps.
    1. Let |config| be |mapping|'s [=fenced frame config mapping/finalized config mapping=][|urn|].
```

gtanzer replied:
It's not an issue for PA because the pending urn is never returned to script, so it can't be passed to deprecatedURNtoTURL; instead the promise waits for the auction result to be resolved. In selectURL it immediately resolves to a pending urn, so you could pass it as input (and it won't work).

JensenPaul replied:
So what's the right thing to put in the spec?  it sounds like it only works with finalized URNs, in which case I think my suggestion in my last comment is accurate?

gtanzer replied:
Yeah, afaict what you put in the comment matches the implementation.

JensenPaul replied:
Okay, I switched to my suggested steps.

domfarolino commented on spec.bs:
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
+        [=resolve=] |p| with |config|'s [=fenced frame config/mapped url=]'s [=mapped url/value=].
+
+     1. If |send_reports| is true, run steps in |config|'s
+        [=fenced frame config/on navigate callback=].
```
Why run these just on lookup? I assume that's what the imple does? It just feels a little strange.

JensenPaul replied:
`deprecatedURNtoURL()` was added eons ago and the motivation was stated in https://github.com/WICG/turtledove/blob/main/Proposed_First_FLEDGE_OT_Details.md#advertisement-rendering as: "the `sendReports` parameter...is included to [promote accurate measurement](https://github.com/WICG/turtledove/blob/main/Proposed_First_FLEDGE_OT_Details.md#promoting-accurate-measurement-of-fledge-effectiveness.) and comparison of FLEDGE API utility by allowing auction winning URLs to be treated and used as they are in today’s ad auctions."
Basically we wanted a way to test Protected Audience (especially reporting scripts and APIs) when the URL coming from `deprecatedURNtoURL()` was rendered in a some other way that's not visible to the browser (so the browser doesn't know when to trigger Protected Audience reporting without the hint from `sendReports`).

domfarolino replied:
Any chance there's a WPT for this?

JensenPaul replied:
It looks like we don't have WPTs for when `sendReports` is `true`, but I should note that `deprecatedURNtoURL()` is [going away at 3PCD](https://developers.google.com/privacy-sandbox/relevance/protected-audience-api/feature-status#:~:text=Note%3A%20The%20temporary%20navigator.deprecatedURNToURL()%20will%20be%20removed%20by%20third%2Dparty%20cookie%20deprecation.).

domfarolino reviewed (commented):
Pretty much lgtm; nothing major here.

JensenPaul commented on spec.bs:
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
+
+  1. Run the following steps [=in parallel=]:
+
+     1. Let |mapping| be |global|'s [=associated Document=]'s [=node navigable=]'s
```
I think I fixed this.  Please take a look.

domfarolino commented on spec.bs:
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
 
+<div algorithm>
+  The <dfn method for=Navigator>deprecatedURNtoURL(|urnOrConfig|, |send_reports|)</dfn>
+  method steps are:
```
All of this can be wrapped like so:

```
  The <dfn method for=Navigator>deprecatedURNtoURL(|urnOrConfig|, |send_reports|)</dfn> method
  steps are:
```

and be compliant with 100 lines.

domfarolino commented on spec.bs:
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
+
+  1. Otherwise, set |urn| to |urnOrConfig|'s [=fencedframeconfig/urn=].
+
+  1. If |urn| is <span class=XXX>TODO invalid</span>, then [=exception/throw=] a {{TypeError}}.
```
https://github.com/WICG/fenced-frame/pull/161#discussion_r1626061123 got marked outdated, so I guess I'll re-file this comment as a stand-in for it.

JensenPaul replied:
I proposed some new words here, WDYT?

domfarolino replied:
I think that sounds right? Although it would be inelegant to have to write that `i.e.,` each time we reference the concept of an invalid URN. But I suppose this is an incremental improvement so that's fine.

domfarolino reviewed (commented):
Btw what is the motivation to include this API in this spec, instead of the PA spec? Impl-wise, it looks like it was included closer to the PA stuff: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ad_auction/navigator_auction.idl;l=41;drc=ac83a5a2d3c04763d86ce16d92f3904cc9566d3a;bpv=1;bpt=0

JensenPaul commented:
> Btw what is the motivation to include this API in this spec, instead of the PA spec? Impl-wise, it looks like it was included closer to the PA stuff: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ad_auction/navigator_auction.idl;l=41;drc=ac83a5a2d3c04763d86ce16d92f3904cc9566d3a;bpv=1;bpt=0

I feel like stuff that only relates to URNs, like `deprecatedURNtoURL()`, should be in the FF spec, like `deprecatedReplaceInURN()` already is.  The use of URNs (and then FFConfig) seems tied to seeding FFs, and could be used by specs other than PA.

domfarolino commented:
Ah darn it looks there are conflicts now that the other PR merged. Would you be able to sort them out please?

JensenPaul commented:
> Ah darn it looks there are conflicts now that the other PR merged. Would you be able to sort them out please?

I fixed the merge conflict.  Now it looks like the build is failing on an unrelated line:
`LINE 1724: Couldn't find WPT test 'fenced-frame/fence-report-event-cross-origin.https.html' - did you misspell something?`
Looks related to https://chromium-review.googlesource.com/c/chromium/src/+/5544804
So I updated the WPT file name.

domfarolino commented on spec.bs:
```diff
@@ -1507,7 +1516,47 @@ partial interface Navigator {
+
+  1. Otherwise, set |urn| to |urnOrConfig|'s [=fencedframeconfig/urn=].
 
+  1. If |urn| is not a valid [=urn uuid=] (i.e. won't pass the ABNF in Section 3 of
```
```suggestion
  1. If |urn| is not a valid [=urn uuid=] (i.e., won't pass the ABNF in Section 3 of
```

nit: https://whatwg.org/style-guide#grammar

domfarolino reviewed (approved):
(Meant to "approve" with nits, whoops)
